hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java

```diff
@@ -84,40 +85,28 @@ public PipelinePlacementPolicy(final NodeManager nodeManager,
     this.heavyNodeCriteria = dnLimit == null ? 0 : Integer.parseInt(dnLimit);
   }
 
-  int currentPipelineCount(DatanodeDetails datanodeDetails, int nodesRequired) {
-
-    // Datanodes from pipeline in some states can also be considered available
-    // for pipeline allocation. Thus the number of these pipeline shall be
-    // deducted from total heaviness calculation.
-    int pipelineNumDeductable = 0;
-    Set<PipelineID> pipelines = nodeManager.getPipelines(datanodeDetails);
-    for (PipelineID pid : pipelines) {
-      Pipeline pipeline;
-      try {
-        pipeline = stateManager.getPipeline(pid);
-      } catch (PipelineNotFoundException e) {
-        LOG.debug("Pipeline not found in pipeline state manager during" +
-            " pipeline creation. PipelineID: {}", pid, e);
-        continue;
-      }
-      if (pipeline != null &&
-          // single node pipeline are not accounted for while determining
-          // the pipeline limit for dn
-          pipeline.getType() == HddsProtos.ReplicationType.RATIS &&
-          (RatisReplicationConfig
-              .hasFactor(pipeline.getReplicationConfig(), ReplicationFactor.ONE)
-              ||
-              pipeline.getReplicationConfig().getRequiredNodes()
-                  == nodesRequired &&
-                  pipeline.getPipelineState()
-                      == Pipeline.PipelineState.CLOSED)) {
-        pipelineNumDeductable++;
-      }
-    }
-    return pipelines.size() - pipelineNumDeductable;
+  int currentRatisThreePipelineCount(DatanodeDetails datanodeDetails) {
+    // Safe to cast collection's size to int
+    return (int) nodeManager.getPipelines(datanodeDetails).stream()
+        .map(id -> {
+          try {
+            return stateManager.getPipeline(id);
+          } catch (PipelineNotFoundException e) {
+            LOG.debug("Pipeline not found in pipeline state manager during" +
+                " pipeline creation. PipelineID: {}", id, e);
+            return null;
+          }
+        })
+        .filter(this::isNonClosedRatisThreePipeline)
+        .count();
   }
 
-
+  private boolean isNonClosedRatisThreePipeline(Pipeline p) {
+    return p.getType() == RATIS
```

Comment:
I think this could be simplified to:

```
p.getReplicationConfig().equals(new RatisReplicationConfig(THREE)) && !p.isClosed()
```

And you could create a `private static final RATIS_THREE_CONFIG = new RatisReplicationConfig(THREE)` and use the constant in the above check to avoid creating a new object each time.

Reply:
> And you could create a `private static final RATIS_THREE_CONFIG = new RatisReplicationConfig(THREE)` and use the constant in the above check to avoid creating a new object each time.

I would like to suggest creating this as `public` in `RatisReplicationConfig`.  This would benefit other parts of Ozone code as well.

Reply:
Yea, I was just looking at the code to see where else we could benefit from this. I think it would be a good idea to a constant in RatisReplicationConfig for ONE and THREE. We could have a separate Jira to move any existing calls to `new RatisReplicationConfig...` over to the constant in non-test code, which might save some allocations.

I wonder if we would be better with a static method in RatisReplicationConfig, perhaps `getInstance(Factor factor)` - right now every ContainerInfo object in SCM holds a unique ReplicationConfig object - if we change the code correctly we could reduce all those duplicated instances down to 1. The object only holds a single field, but it will still make a small difference to memory.

I will create a followup Jira for this purpose and then we could implement that part separately to avoid blocking this one.

Reply:
> will create a followup Jira

Sure, I wasn't suggesting to perform all that refactoring in this PR.

Reply:
I went ahead and did the refactor in #3165 - the diff is large but repetitive. If we can commit it, then we could do the small refactor here. Lets see if the CI goes green in #3165 to start with!

Reply:
Thanks for the comments @sodonnel @adoroszlai, it would be simpler that way. Thanks for the followup Jira, I'll wait until the #3165 is committed.

Reply:
#3165 is committed now, so please go ahead and make the changes here.